lib/api/src/externals/function.rs

```diff
@@ -651,16 +673,14 @@ impl Function {
         }
     }
 
-    pub(crate) fn checked_anyfunc(&self) -> VMCallerCheckedAnyfunc {
-        let vmsignature = self
-            .store
-            .engine()
-            .register_signature(&self.exported.vm_function.signature);
-        VMCallerCheckedAnyfunc {
+    pub(crate) fn checked_anyfunc(&self) -> VMFuncRef {
+        let engine = self.store.engine();
+        let vmsignature = engine.register_signature(&self.exported.vm_function.signature);
+        engine.register_function_metadata(VMCallerCheckedAnyfunc {
```

Comment:
I'm not super sure about this model. (the engine keeping track of the metadata) as a parent -> children relation.
I need to think more about this

Reply:
Yeah, the specific issue that this solves is the problem of arbitrary functions inserted into tables. Without that issue there are simpler solutions.

Explicitly, the problem is that users at runtime can get access to an exported or imported table, and insert any arbitrary function into it. Including functions that have signatures that do not show up in either the exported or imported functions of that module (we also handle that case with `engine.register_signature`). Thus we need a `FuncRef` (which is a pointer to data somewhere else) for any possible function a user can create at runtime, additionally we have to store that data somewhere that the `FuncRef` points to.